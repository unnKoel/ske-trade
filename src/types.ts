export enum Trend {
  Up = 'Up',
  Down = 'Down',
  Origin = 'Origin',
}

export enum TradeStatus {
  New = 'New',
  Updated = 'Updated',
  Closed = 'Closed',
}

export interface TradeEntity {
  tradeId: string
  tradeName: string
  tradeSymbol: string
  currentPrice: number
  lastPrice: number
  traderName: string
  trend: Trend
  updateTime: string
  createTime: string
  tradeStatus: TradeStatus
  updated?: boolean
  created?: boolean
  highlight?: boolean
}

export enum ResponseCode {
  Success = 1,
  Failure = 0,
}

export interface Response<T = unknown> {
  eventType: string
  code: number
  data: T
}

export interface GetTradesAPIReturn {
  totalAmount: number
  trades: TradeEntity[]
  hasMore: boolean
}

export interface UpdateTradesAPIReturn {
  addTrades: TradeEntity[]
  updateTrades: TradeEntity[]
  totalAmount: number
}

export interface DeleteTradeAPIReturn {
  totalAmount: number
}

type RGB = `rgb(${number}, ${number}, ${number})`
type RGBA = `rgba(${number}, ${number}, ${number}, ${number})`
type HEX = `#${string}`

export type Color = RGB | RGBA | HEX
