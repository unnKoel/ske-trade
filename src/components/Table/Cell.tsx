import { Cell as ReactTableCell } from 'react-table'
import styled from 'styled-components'
import { Trend } from '../../types'
import ArrowUp from './up.svg'
import ArrowDown from './down.svg'
import Horizontal from './horizontal.svg'

interface CellProps {
  cellDescription: ReactTableCell<object, any>
}

console.log('ArrowUp', ArrowUp)

const StyledArrowUp = styled(ArrowUp)`
  width: 8em;
  height: 2em;
`
const StyledArrowDown = styled(ArrowDown)`
  width: 8em;
  height: 2em;
`
const StyledHorizontal = styled(Horizontal)`
  width: 8em;
  height: 2em;
`

const Cell = ({ cellDescription }: CellProps) => {
  const {
    column: { id },
    value,
  } = cellDescription

  let CellNode = null

  // As trend field, disply arrow icon for different value
  if (id === 'trend') {
    CellNode =
      value === Trend.Up ? (
        <StyledArrowUp />
      ) : value === Trend.Down ? (
        <StyledArrowDown />
      ) : (
        <StyledHorizontal />
      )
  } else {
    CellNode = cellDescription.render('Cell')
  }

  return (
    <div {...cellDescription.getCellProps()} className='td'>
      {CellNode}
    </div>
  )
}

export default Cell
