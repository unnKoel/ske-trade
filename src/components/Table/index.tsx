import { Column, useBlockLayout, useTable } from 'react-table'
import { useMemo, useCallback, ComponentType, CSSProperties } from 'react'
import { FixedSizeList, ListChildComponentProps } from 'react-window'
import styled from 'styled-components'

interface TableProp {
  columns: Array<Column<object>>
  data: object[]
}

const Styles = styled.div`
  padding: 1rem;

  .table {
    display: inline-block;
    border-spacing: 0;
    border: 1px solid black;

    .tr {
      :last-child {
        .td {
          border-bottom: 0;
        }
      }
    }

    .th,
    .td {
      margin: 0;
      padding: 0.5rem;
      border-bottom: 1px solid black;
      border-right: 1px solid black;

      :last-child {
        border-right: 1px solid black;
      }
    }
  }
`

const Table = ({ columns, data }: TableProp): JSX.Element => {
  const scrollBarSize = useMemo(() => scrollbarWidth(), [])

  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    totalColumnsWidth,
    prepareRow,
  } = useTable(
    {
      columns,
      data,
    },
    useBlockLayout,
  )

  const RenderRow = useCallback(
    ({ index, style }: { index: number; style: CSSProperties }) => {
      const row = rows[index]
      prepareRow(row)
      return (
        <div
          {...row.getRowProps({
            style,
          })}
          className='tr'
        >
          {row.cells.map((cell) => {
            return (
              // eslint-disable-next-line react/jsx-key
              <div {...cell.getCellProps()} className='td'>
                {cell.render('Cell')}
              </div>
            )
          })}
        </div>
      )
    },
    [prepareRow, rows],
  )

  return (
    <Styles>
      <div {...getTableProps()} className='table'>
        <div>
          {headerGroups.map((headerGroup) => (
            // eslint-disable-next-line react/jsx-key
            <div {...headerGroup.getHeaderGroupProps()} className='tr'>
              {headerGroup.headers.map((column) => (
                // eslint-disable-next-line react/jsx-key
                <div {...column.getHeaderProps()} className='th'>
                  {column.render('Header')}
                </div>
              ))}
            </div>
          ))}
        </div>

        <div {...getTableBodyProps()}>
          <FixedSizeList
            height={400}
            itemCount={rows.length}
            itemSize={35}
            width={totalColumnsWidth + scrollBarSize}
          >
            {RenderRow as ComponentType<ListChildComponentProps<any>>}
          </FixedSizeList>
        </div>
      </div>
    </Styles>
  )
}

const scrollbarWidth = (): number => {
  // https://davidwalsh.name/detect-scrollbar-width
  const scrollDiv = document.createElement('div')
  scrollDiv.setAttribute(
    'style',
    'width: 100px; height: 100px; overflow: scroll; position:absolute; top:-9999px;',
  )
  document.body.appendChild(scrollDiv)
  const scrollbarWidth = scrollDiv.offsetWidth - scrollDiv.clientWidth
  document.body.removeChild(scrollDiv)
  return scrollbarWidth
}

export default Table
