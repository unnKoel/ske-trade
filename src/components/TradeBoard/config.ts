export const columns = [
  {
    Header: 'Trade Name',
    accessor: 'tradeName',
  },
  {
    Header: 'Trade Symbol',
    accessor: 'tradeSymbol',
  },
  {
    Header: 'Current Price',
    accessor: 'currentPrice',
  },
  {
    Header: 'Last Price',
    accessor: 'lastPrice',
  },
  {
    Header: 'Trader Name',
    accessor: 'traderName',
    width: 200,
  },
  {
    Header: 'Trend',
    accessor: 'trend',
  },
  {
    Header: 'Update Time',
    accessor: 'updateTime',
    width: 180,
  },
  {
    Header: 'Create Time',
    accessor: 'createTime',
    width: 180,
  },
]
