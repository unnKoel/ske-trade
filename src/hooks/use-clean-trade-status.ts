import { createContext, useContext } from 'react'

const creactCleanTradeStatus = () => {
  const CleanTradeStatusContext = createContext<
    (tradeId: string, properties: Record<'updated' | 'created', false>) => void
  >(() => {})

  const useCleanTradeStatus = () => {
    const cleanTradeStatus = useContext(CleanTradeStatusContext)

    return cleanTradeStatus
  }

  return {
    useCleanTradeStatus,
    CleanTradeStatusProvider: CleanTradeStatusContext.Provider,
  }
}

const { useCleanTradeStatus, CleanTradeStatusProvider } =
  creactCleanTradeStatus()

export { useCleanTradeStatus, CleanTradeStatusProvider }
