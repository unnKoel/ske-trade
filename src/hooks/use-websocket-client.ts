import { createContext, useContext } from 'react'
import { WebSocketClient } from '../websocket-client'

const creactWebSocketClient = () => {
  const WebsocketClientContext = createContext<WebSocketClient | null>(null)

  const useWebSocketClient = () => {
    const websocketClient = useContext(WebsocketClientContext)

    return websocketClient
  }

  return {
    useWebSocketClient,
    WebsocketClientProvider: WebsocketClientContext.Provider,
  }
}

const { useWebSocketClient, WebsocketClientProvider } = creactWebSocketClient()

export { useWebSocketClient, WebsocketClientProvider }
