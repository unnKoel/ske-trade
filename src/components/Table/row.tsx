import { useEffect, useState } from 'react'
import { timer } from 'rxjs'
import { Row as ReactTableRow, TableRowProps } from 'react-table'
import styled from 'styled-components'
import { useCleanTradeStatus } from '../../hooks/use-clean-trade-status'
import { Color, TradeEntity, Trend } from '../../types'
import Cell from './Cell'

export interface RowColors {
  up: Color
  down: Color
  same: Color
  highlight: Color
  new: Color
}

export interface RowProps {
  row: ReactTableRow<object>
  tableRowProps: TableRowProps
  rowColors?: RowColors
}

const REGULAR_COLOR = '#fff' as Color

interface RowWrapperProps {
  readonly backgroundColor: Color
}
const RowWrapper = styled.div<RowWrapperProps>`
  background: ${(props) => props.backgroundColor};
`

const Row = ({
  row,
  tableRowProps,
  rowColors = {
    up: '#FF0000',
    down: '#008000',
    same: '#808080',
    highlight: '#0000FF',
    new: '#FFFF00',
  },
}: RowProps) => {
  const [backgroundColor, setBackgroundColor] = useState(REGULAR_COLOR)
  const cleanTradeStatus = useCleanTradeStatus()

  useEffect(function () {
    const { updated, created, tradeId } = row.original as TradeEntity
    const { trend } = row.values
    let backgroundColor = REGULAR_COLOR

    if (updated) {
      backgroundColor =
        trend === Trend.Up
          ? rowColors.up
          : trend === Trend.Down
          ? rowColors.down
          : rowColors.same
    }

    if (created) {
      backgroundColor = rowColors.new
    }

    setBackgroundColor(backgroundColor)
    timer(500).subscribe(() => {
      setBackgroundColor(REGULAR_COLOR)
      cleanTradeStatus(tradeId, {
        updated: false,
        created: false,
      })
    })
  }, [])

  return (
    <RowWrapper
      {...tableRowProps}
      className='tr'
      backgroundColor={backgroundColor}
    >
      {row.cells.map((cell) => (
        <Cell cellDescription={cell} key={cell.column.id}></Cell>
      ))}
    </RowWrapper>
  )
}

export default Row
