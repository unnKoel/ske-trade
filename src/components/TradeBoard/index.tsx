import { useCallback, useEffect, useMemo, useState } from 'react'
import { firstValueFrom } from 'rxjs'
import { filter } from 'rxjs/operators'
import styled from 'styled-components'
import { produce } from 'immer'
import moment from 'moment'
import { useWebSocketClient } from '../../hooks/use-websocket-client'
import Table from '../Table'
import { columns } from './config'
import {
  TradeEntity,
  GetTradesAPIReturn,
  UpdateTradesAPIReturn,
} from '../../types'
import { CleanTradeStatusProvider } from '../../hooks/use-clean-trade-status'

const Title = styled.h2`
  padding-left: 1rem;
`

const TradeBoard = (): JSX.Element => {
  const [tradeMap, setTradeMap] = useState<Record<string, TradeEntity>>({})
  const [hasNextPage, setHasNextPage] = useState(true)
  const [page, setPage] = useState(1)
  const [isNextPageLoading, setIsNextPageLoading] = useState(false)
  const webSocketClient = useWebSocketClient()

  const loadTrades = useCallback(
    async (...args: any) => {
      setIsNextPageLoading(true)
      const response = await firstValueFrom(
        webSocketClient?.request<GetTradesAPIReturn>({
          eventType: 'getTrades',
          page,
          amount: 10,
        }),
      )
      const { trades, hasMore } = response.data
      const trade = trades.reduce<Record<string, TradeEntity>>((acc, trade) => {
        acc[trade.tradeId] = convertTradeFormat(trade)
        return acc
      }, {})
      setTradeMap({ ...tradeMap, ...trade })
      setIsNextPageLoading(false)
      setHasNextPage(hasMore)
      setPage(page + 1)
    },
    [tradeMap, page],
  )

  /** */
  useEffect(function observeTradesChange() {
    const subscription = webSocketClient?.responseObservable
      .pipe(filter(({ eventType }) => eventType === 'updateTrades'))
      .subscribe((response) => {
        const data = response.data as UpdateTradesAPIReturn
        const { updateTrades, addTrades } = data
        setTradeMap(
          produce((draftState) => {
            updateTrades.forEach((updateTrade) => {
              draftState[updateTrade.tradeId] = markUpdated(
                convertTradeFormat(updateTrade),
              )
            })
            addTrades.forEach((addTrade) => {
              draftState[addTrade.tradeId] = markCreated(
                convertTradeFormat(addTrade),
              )
            })
          }),
        )
      })

    return () => subscription.unsubscribe()
  }, [])

  const cleanTradeStatus = useCallback(
    (tradeId: string, properties: Record<'updated' | 'created', false>) => {
      setTradeMap(
        produce((draftState) => {
          draftState[tradeId].updated = properties.updated
          draftState[tradeId].created = properties.created

          return draftState
        }),
      )
    },
    [],
  )

  const trades = useMemo(() => Object.values(tradeMap), [tradeMap])

  console.log('trades', trades.length)

  return (
    <CleanTradeStatusProvider value={cleanTradeStatus}>
      <Title>Trade Board Table</Title>
      <Table
        columns={columns}
        data={trades}
        isNextPageLoading={isNextPageLoading}
        loadMore={loadTrades}
        hasNextPage={hasNextPage}
      />
    </CleanTradeStatusProvider>
  )
}

const convertTradeFormat = (trade: TradeEntity) => {
  return {
    ...trade,
    updateTime: moment(trade.updateTime).format('hh:mm:ss MM/DD/YYYY'),
    createTime: moment(trade.createTime).format('hh:mm:ss MM/DD/YYYY'),
    updated: false,
    created: false,
  }
}

const markUpdated = (trade: TradeEntity) => {
  return {
    ...trade,
    updated: true,
  }
}

const markCreated = (trade: TradeEntity) => {
  return {
    ...trade,
    created: true,
  }
}

export default TradeBoard
