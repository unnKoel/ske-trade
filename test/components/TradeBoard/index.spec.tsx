import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import { BehaviorSubject } from 'rxjs'
import TradeBoard from '@app/components/TradeBoard'
import { WebsocketClientProvider } from '@app/hooks/use-websocket-client'
import createWebSocketClient, { WebSocketClient } from '@app/websocket-client'
import { Response } from '@app/types'
import Table from '@app/components/Table'
import mockData from './data.json'

export interface MockWebSocketClient extends WebSocketClient {
  send: (data: Response) => void
}

const mockResponseSubject = new BehaviorSubject<Response>(mockData)
const mockResponseObservable = mockResponseSubject.asObservable()
jest.mock('@app/websocket-client', () => {
  return {
    __esModule: true,
    default: () => {
      return {
        send: (data: Response) => mockResponseSubject.next(data),
        request: jest.fn(() => {
          return mockResponseObservable
        }),
        responseObservable: mockResponseObservable,
      }
    },
  }
})

const webSocketClient = createWebSocketClient('') as MockWebSocketClient

it('render TradeBoard', async () => {
  render(
    <WebsocketClientProvider value={webSocketClient}>
      <TradeBoard />
    </WebsocketClientProvider>,
  )

  expect(await screen.findAllByText('BTC/USDT')).toHaveLength(4)
})

it.skip('render TradeBoard with trades returned', () => {
  const component = renderer.create(
    <WebsocketClientProvider value={webSocketClient}>
      <TradeBoard />
    </WebsocketClientProvider>,
  )

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  renderer.act(() => {
    component.root.findAllByType(Table)[0].props.loadMore()
  })

  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
