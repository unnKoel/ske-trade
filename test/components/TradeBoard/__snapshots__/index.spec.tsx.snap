// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render TradeBoard with trades returned 1`] = `
[
  <h2
    className="sc-gLDzan buDzqq"
  >
    Trade Board Table
  </h2>,
  <div
    className="sc-jTrPJq fRLbar"
  >
    <div
      className="table"
      role="table"
    >
      <div>
        <div
          className="tr"
          role="row"
          style={
            {
              "display": "flex",
              "width": "1350px",
            }
          }
        >
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "150px",
              }
            }
          >
            Trade Name
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "150px",
              }
            }
          >
            Trade Symbol
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "150px",
              }
            }
          >
            Current Price
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "150px",
              }
            }
          >
            Last Price
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "200px",
              }
            }
          >
            Trader Name
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "150px",
              }
            }
          >
            Trend
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "200px",
              }
            }
          >
            Update Time
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "200px",
              }
            }
          >
            Create Time
          </div>
        </div>
      </div>
      <div
        onContextMenu={[Function]}
        role="rowgroup"
      >
        <div
          onScroll={[Function]}
          style={
            {
              "WebkitOverflowScrolling": "touch",
              "direction": "ltr",
              "height": 500,
              "overflow": "auto",
              "position": "relative",
              "width": 1350,
              "willChange": "transform",
            }
          }
        >
          <div
            style={
              {
                "height": 40,
                "pointerEvents": undefined,
                "width": "100%",
              }
            }
          >
            <div
              style={
                {
                  "height": 40,
                  "left": 0,
                  "position": "absolute",
                  "right": undefined,
                  "top": 0,
                  "width": "100%",
                }
              }
            >
              Loading
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-eDDNvR QpTv"
      open={false}
    >
      <ul>
        <li
          onClick={[Function]}
        >
          remove
        </li>
        <li
          onClick={[Function]}
        >
          highlight
        </li>
      </ul>
    </div>
  </div>,
  <div
    className="sc-iAEyYk jShygk"
  >
    Total amount of trades: 
    0
  </div>,
]
`;

exports[`render TradeBoard with trades returned 2`] = `
[
  <h2
    className="sc-gLDzan buDzqq"
  >
    Trade Board Table
  </h2>,
  <div
    className="sc-jTrPJq fRLbar"
  >
    <div
      className="table"
      role="table"
    >
      <div>
        <div
          className="tr"
          role="row"
          style={
            {
              "display": "flex",
              "width": "1350px",
            }
          }
        >
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "150px",
              }
            }
          >
            Trade Name
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "150px",
              }
            }
          >
            Trade Symbol
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "150px",
              }
            }
          >
            Current Price
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "150px",
              }
            }
          >
            Last Price
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "200px",
              }
            }
          >
            Trader Name
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "150px",
              }
            }
          >
            Trend
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "200px",
              }
            }
          >
            Update Time
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              {
                "boxSizing": "border-box",
                "display": "inline-block",
                "width": "200px",
              }
            }
          >
            Create Time
          </div>
        </div>
      </div>
      <div
        onContextMenu={[Function]}
        role="rowgroup"
      >
        <div
          onScroll={[Function]}
          style={
            {
              "WebkitOverflowScrolling": "touch",
              "direction": "ltr",
              "height": 500,
              "overflow": "auto",
              "position": "relative",
              "width": 1350,
              "willChange": "transform",
            }
          }
        >
          <div
            style={
              {
                "height": 40,
                "pointerEvents": undefined,
                "width": "100%",
              }
            }
          >
            <div
              style={
                {
                  "height": 40,
                  "left": 0,
                  "position": "absolute",
                  "right": undefined,
                  "top": 0,
                  "width": "100%",
                }
              }
            >
              Loading
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-eDDNvR QpTv"
      open={false}
    >
      <ul>
        <li
          onClick={[Function]}
        >
          remove
        </li>
        <li
          onClick={[Function]}
        >
          highlight
        </li>
      </ul>
    </div>
  </div>,
  <div
    className="sc-iAEyYk jShygk"
  >
    Total amount of trades: 
    0
  </div>,
]
`;
